src/vos/vos_dtx.c

```diff
@@ -102,179 +171,246 @@ dtx_hkey_cmp(struct btr_instance *tins, struct btr_record *rec, void *hkey)
+					sizeof(dae_df->dae_flags));
+			dae_df->dae_flags = DTX_EF_INVALID;
+
+			D_ASSERT(dsb->dsb_count > bcb->bcb_dae_count);
+			umem_tx_add_ptr(umm, &dsb->dsb_count,
+					sizeof(dsb->dsb_count));
+			dsb->dsb_count--;
+
+			if (!dtx_is_null(bcb->bcb_recs[idx])) {
+				umem_free(umm, bcb->bcb_recs[idx]);
+				bcb->bcb_recs[idx] = UMOFF_NULL;
+			}
+
+			rc = vos_tx_end(umm, 0);
+		}
+
+		d_list_del(&dce->dce_bcb_link);
 	}
 
-	dbtree_close(hdl);
+	D_FREE_PTR(dce);
 
-	rc = dbtree_create_inplace(VOS_BTR_DTX_TABLE, 0,
-				   DTX_BTREE_ORDER, &pool->vp_uma,
-				   &dtab_df->tt_committed_btr, &hdl);
-	if (rc != 0) {
-		D_ERROR("Failed to create DTX committed dbtree for pool "
-			DF_UUID": rc = %d\n", DP_UUID(pool->vp_id), rc);
-		return rc;
+	return rc;
+}
+
+static int
+dtx_cmt_ent_fetch(struct btr_instance *tins, struct btr_record *rec,
+		  d_iov_t *key_iov, d_iov_t *val_iov)
+{
+	if (val_iov != NULL) {
+		struct vos_dtx_cmt_ent	*dce;
+
+		dce = umem_off2ptr(&tins->ti_umm, rec->rec_off);
+		d_iov_set(val_iov, dce, sizeof(*dce));
 	}
 
-	dtab_df->tt_count = 0;
-	dtab_df->tt_entry_head = UMOFF_NULL;
-	dtab_df->tt_entry_tail = UMOFF_NULL;
+	return 0;
+}
+
+static int
+dtx_cmt_ent_update(struct btr_instance *tins, struct btr_record *rec,
+		   d_iov_t *key, d_iov_t *val)
+{
+	struct vos_dtx_cmt_ent	*dce = val->iov_buf;
+
+	dce->dce_exist = 1;
 
-	dbtree_close(hdl);
 	return 0;
 }
 
+static btr_ops_t dtx_committed_btr_ops = {
+	.to_hkey_size	= dtx_hkey_size,
+	.to_hkey_gen	= dtx_hkey_gen,
+	.to_hkey_cmp	= dtx_hkey_cmp,
+	.to_rec_alloc	= dtx_cmt_ent_alloc,
+	.to_rec_free	= dtx_cmt_ent_free,
+	.to_rec_fetch	= dtx_cmt_ent_fetch,
+	.to_rec_update	= dtx_cmt_ent_update,
+};
+
 int
-vos_dtx_table_destroy(struct vos_pool *pool, struct vos_dtx_table_df *dtab_df)
+vos_dtx_table_register(void)
 {
-	daos_handle_t	hdl;
-	int		rc = 0;
+	int	rc;
 
-	if (pool == NULL || dtab_df == NULL) {
-		D_ERROR("Invalid handle\n");
-		return -DER_INVAL;
+	rc = dbtree_class_register(VOS_BTR_DTX_ACT_TABLE, 0,
+				   &dtx_active_btr_ops);
+	if (rc != 0) {
+		D_ERROR("Failed to register DTX active dbtree: %d\n", rc);
+		return rc;
 	}
 
-	if (dtab_df->tt_active_btr.tr_class != 0) {
-		rc = dbtree_open_inplace(&dtab_df->tt_active_btr,
-					 &pool->vp_uma, &hdl);
-		if (rc == 0)
-			rc = dbtree_destroy(hdl, NULL);
+	rc = dbtree_class_register(VOS_BTR_DTX_CMT_TABLE, 0,
+				   &dtx_committed_btr_ops);
+	if (rc != 0)
+		D_ERROR("Failed to register DTX committed dbtree: %d\n", rc);
 
-		if (rc != 0)
-			D_ERROR("Fail to destroy DTX active dbtree for pool"
-				DF_UUID": rc = %d\n", DP_UUID(pool->vp_id), rc);
+	return rc;
+}
+
+void
+vos_dtx_table_destroy(struct umem_instance *umm, struct vos_cont_df *cont_df)
+{
+	struct vos_dtx_scm_blob		*dsb;
+	umem_off_t			 dsb_off;
+
+	while (!dtx_is_null(cont_df->cd_dtx_committed_head)) {
+		dsb_off = cont_df->cd_dtx_committed_head;
+		dsb = umem_off2ptr(umm, dsb_off);
+		cont_df->cd_dtx_committed_head = dsb->dsb_next;
```

Comment:
even you are going to destroy the container, but if you change this pionter, then please call tx_add_range otherwise TX abort might get corrupted container

Reply:
OK